src/canopy/llm/cohere.py

```diff
@@ -0,0 +1,325 @@
+        self._client = cohere.Client(api_key, api_url=base_url)
+        self.default_model_params = kwargs
+
+    def chat_completion(self,
+                        system_prompt: str,
+                        chat_history: Messages,
+                        context: Optional[Context] = None,
+                        *,
+                        stream: bool = False,
+                        max_tokens: Optional[int] = None,
+                        model_params: Optional[dict] = None,
+                        ) -> Union[ChatResponse, Iterable[StreamingChatChunk]]:
+        """
+        Chat completion using the Cohere API.
+
+        Note: this function is wrapped in a retry decorator to handle transient errors.
+
+        Args:
+            system_prompt: The system prompt to use for the chat completion (preamble).
+            chat_history: Messages (chat history) to send to the model.
+            context: Knowledge base context to use for the chat completion. Defaults to None (no context).
+            stream: Whether to stream the response or not.
+            max_tokens: Maximum number of tokens to generate. Defaults to None (generates until stop sequence or until hitting max context size).
+            model_params: Model parameters to use for this request. Defaults to None (uses the default model parameters).
+                          Dictonary of parametrs to override the default model parameters if set on initialization.
+                          For example, you can pass: {"temperature": 0.9, "top_p": 1.0} to override the default temperature and top_p.
+                          see: https://platform.openai.com/docs/api-reference/chat/create
+        Returns:
+            ChatResponse or StreamingChatChunk
+
+        Usage:
+            >>> from canopy.llm import OpenAILLM
+            >>> from canopy.models.data_models import UserMessage
+            >>> llm = CohereLLM()
+            >>> messages = [UserMessage(content="Hello! How are you?")]
+            >>> result = llm.chat_completion(messages)
+            >>> print(result.choices[0].message.content)
+            "I'm good, how are you?"
+        """  # noqa: E501
+        model_params_dict: Dict[str, Any] = deepcopy(self.default_model_params)
+        model_params_dict.update(
+            model_params or {}
+        )
+        model_params_dict["max_tokens"] = max_tokens
+
+        model_params_dict = self._convert_model_params(model_params_dict)
+
+        connectors = model_params_dict.pop('connectors', None)
+        messages: List[Dict[str, Any]] = self._map_messages(chat_history)
+
+        if not messages:
+            raise cohere.error.CohereAPIError("No message provided")
+
+        response = self._client.chat(
+            model=self.model_name,
+            message=messages.pop()['message'],
+            chat_history=messages,
+            documents=self.generate_documents_from_context(context),
+            preamble_override=system_prompt,
```

Comment:
By default, this should actually be empty since Cohere handles the system prompt for you as long as the documents parameter is used. So perhaps we can replace "system_prompt" with "preamble_override" as a Cohere-only supported parameter?

Reply:
See my comments above. The `LLM` class shouldn't assume the use case it is being used for (e.g. RAG)